### ISSUE TITLE
Feature request: forward exit code

### ISSUE THREAD

**Issue by fracting** [mintty/mintty#510]
Greetings!

Thanks for the great work on mintty.

I'd like to proposal a feature request: forward exit code of child process to Mintty.

For example, if I run `mintty --forward-exit -e /usr/bin/bash -c exit 3`, I hope the final exit code of mintty itself is also `3`. This is useful for me when running mintty + msys2 / Cygwin on a continuous integration server, I need the exit code to determine if there is anything wrong when running the scripts.

Is that a valid request and also useful for other people?

(Merry Christmas!)

**Comment by ElvenSpellmaker** [comment]
Out of interest why can't you use bash directly instead of proxying it through mintty?

**Comment by fracting** [comment]
Hi, thanks for response.

Due to https://bugs.wine-staging.com/show_bug.cgi?id=290 , Cygwin / MSYS2 doesn't work directly in Linux shell, so we need mintty here.

It is definitely not mintty's fault, in theory there are multiple solutions:
- One solution is using `script.exe` from Cygwin/MSYS2 which supports exit code forwarding, I've been using that solution for a while https://github.com/fracting/wine-fracting/wiki/MSYS2-on-Wine#build-status , but the `script.exe` solution is complicate and not nice, a mintty solution will be much nicer.
- Another solution is implementing logging support in wineconsole, I've record down it as https://bugs.wine-staging.com/show_bug.cgi?id=629
- The last solution is fixing wine staging bug 290, which might be the hardest one since it is still not clear that if it is possible to fix yet.

My main purpose is to use mintty in mainstream CI server like Travis CI / Drone CI / etc. Those CI server does not support Windows, but we've spent some months of work to make Wine support Cygwin / MSYS2. Last time I also tried MSYS2 in Appveyor CI (which has native Windows support), and ran into console related issue, for unknown reason MSYS2 doesn't work directly in Appveyor CI's cmd shell, however if there is an easy way to forward both exit code and console log from mintty to cmd, then these kind of problem could be easily solved.

In summary, I think "forwarding console log" is equal important to "forwarding exit status", if a user want console log forwarding, he might also need exit status forwarding at the same situation.

Cheers!

**Comment by mintty** [comment]
I *think* mintty could provide that feature, maybe even by default. Will have to check. On the other hand, as other terminals do not typically forward an exit code, this solution would be highly specialised and make you dependent on mintty; for that reason, I would rather suggest you find a more portable solution.

**Comment by ElvenSpellmaker** [comment]
As @mintty says you'd be tying your tests to mintty then which might not be a good thing...

The only portable version I know involves using a workaround technique such as:
```bash
exitfile=$(mktest)
/bin/mintty -e bash -c '/command/to/run/here; echo $? >' $exitfile
exitcode=$(<$exitfile)
rm $exitfile
```
If you have a really old shell then this is the *most* portable way (but it's slower):
```bash
exitfile=`mktest`
/bin/mintty -e bash -c '/command/to/run/here; echo $? >' $exitfile
exitcode=`cat $exitfile`
rm $exitfile
```

**Comment by fracting** [comment]
Hey @mintty @ElvenSpellmaker 

Thanks a lot for the kind advice.

I was using a very similar way in my script:
https://github.com/fracting/msys2-scripts/blob/master/src/msys2-shell#L46
(output the exit code to a exit file then cat the exitfile)

Personally I don't mind to be tied by mintty, because I am working on a CI server, and mintty would be part of the CI server by default, rather than part of the software project to test.

I agree that the `exitfile` way is more portable, at the same time I would also appreciate for this feature in mintty :)

Happy new year!

**Comment by mintty** [comment]
On second thought, and check, as no other terminal supports this feature, I do not want to lure people into writing unportable scripts. Which means a design decision against this feature, sorry.

**Comment by fracting** [comment]
Never mind, make sense, thank you all the same.
2016年1月7日 上午1:46，"mintty" <notifications@github.com>写道：
